src/data_types/tensor_common.hpp

```diff
@@ -0,0 +1,459 @@
+// SPDX-License-Identifier: MIT
+#pragma once
+#include <array>
+#include <cassert>
+
+#include <ddc/ddc.hpp>
+
+#include "ddc_aliases.hpp"
+#include "tensor_index_tools.hpp"
+#include "vector_index_tools.hpp"
+
+namespace detail {
+
+template <class ValidIndexSet>
+struct ToCoord;
+
+template <class... Dims>
+struct ToCoord<VectorIndexSet<Dims...>>
+{
+    using type = Coord<Dims...>;
+};
+
+template <class ValidIndexSet>
+using to_coord_t = typename ToCoord<ValidIndexSet>::type;
+
+class LocalExecutionSpace;
+
+template <class T>
+inline constexpr bool enable_tensor_type = false;
+
+} // namespace detail
+
+template <class Type>
+inline constexpr bool is_tensor_type_v
+        = detail::enable_tensor_type<std::remove_const_t<std::remove_reference_t<Type>>>;
+
+/**
+ * @brief A class representing a TensorCommon.
+ * @tparam ElementType The type of the elements of the tensor (usually double/complex).
+ * @tparam ExecutionSpace The space where the code will run. If the execution space is
+ *                  detail::LocalExecutionSpace then the tensor will store its own
+ *                  memory and will be accessible locally (on GPU or CPU depending on
+ *                  the context). Otherwise a memory block must be provided and the
+ *                  tensor will be accessible from the specified execution space.
+ * @tparam ValidIndexSet The indices that can be used along each dimension of the tensor.
+ */
+template <class ElementType, class ExecutionSpace, class... ValidIndexSet>
+class TensorCommon
+{
+    static_assert((is_vector_index_set_v<ValidIndexSet> && ...));
+    static_assert(
+            (((is_covariant_vector_index_set_v<ValidIndexSet>)
+              || (is_contravariant_vector_index_set_v<ValidIndexSet>))
+             && ...));
+
+public:
+    /// The TensorIndexSet describing the possible indices.
+    using index_set = ddc::detail::TypeSeq<ValidIndexSet...>;
+
+protected:
+    /// The number of elements in the mdspan
+    static constexpr std::size_t s_n_elements = (ddc::type_seq_size_v<ValidIndexSet> * ...);
+    /// The type of the Kokkos mdspan layout
+    using layout_type = std::conditional_t<
+            std::is_same_v<ExecutionSpace, detail::LocalExecutionSpace>,
+            Kokkos::layout_right,
+            Kokkos::layout_stride>;
```

Comment:
Why changing the layout bse on the execution space ?

Reply:
My hope is to modify `VectorField` so instead of containing `std::array<FieldType, n_dims>` it contains a Field with an extra dimension. I could then use `[]` to get a (strided) mdspan (via a Field) containing the `Vector`. As a second step I want to add a `TensorRef` inheriting from `TensorCommon` which is non-owning. This would store the mdspan coming from `VectorField` so it can be used in the same way as a `Vector` defined by `Tensor`. The aim is to be able to remove this function:
https://github.com/gyselax/gyselalibxx/blob/c1b2bbb5192a99c66c24064418846eac6dfe61f2/src/data_types/vector_field.hpp#L507-L533

and this const in the return:
https://github.com/gyselax/gyselalibxx/blob/c1b2bbb5192a99c66c24064418846eac6dfe61f2/src/data_types/vector_field.hpp#L361

Reply:
Maybe clearer
8deafa61125f13b63237c513349c5e63c62fae31